### PR TITLE
doc: add docstrings and examples for `String` functions

### PR DESCRIPTION
**Pull request by austinletson**
Add docstrings, usage examples, and doctests for `String.get'`, `String.next'`, `String.posOf`, `String.revPosOf`.

### CODE REVIEW

File: src/Init/Data/String/Basic.lean [hunk 1]
```diff
@@ -256,15 +256,50 @@ def atEnd : (@& String) → (@& Pos) → Bool
...
+
+Examples:
+Given `def abc := "abc"` and `def lean := "L∃∀N"`,
+* `abc.get' 0 (by decide) = 'a'`
+* `lean.get' (0 |> lean.next |> lean.next) (by decide) = '∀'`
+
+A typical pattern combines `get'` with a dependent if-else expression
+to avoid the overhead of an additional bounds check. For example:
+```
+def getInBounds? (s : String) (p: String.Pos) : Option Char :=
```
Comment (semorrison) [inline src/Init/Data/String/Basic.lean:273]:
```suggestion
def getInBounds? (s : String) (p : String.Pos) : Option Char :=
```

File: src/Init/Data/String/Basic.lean [hunk 2]
```diff
@@ -256,15 +256,50 @@ def atEnd : (@& String) → (@& Pos) → Bool
...
+Returns the character at position `p` of a string.
+If `p` is not a valid position, returns `(default : Char)`.
+
+Requires evidence, `h`, that `p` is within bounds
+instead of performing a runtime bounds check as in `get`.
+
+Examples:
+Given `def abc := "abc"` and `def lean := "L∃∀N"`,
+* `abc.get' 0 (by decide) = 'a'`
+* `lean.get' (0 |> lean.next |> lean.next) (by decide) = '∀'`
```
Comment (semorrison) [inline src/Init/Data/String/Basic.lean:268]: May clearer without the defs, and just inlining the strings?

File: src/Init/Data/String/Basic.lean (hunk 1 above)
Comment (semorrison) [reply src/Init/Data/String/Basic.lean:273]: and below

File: src/Init/Data/String/Basic.lean (hunk 2 above)
Comment (semorrison) [reply src/Init/Data/String/Basic.lean:268]: and below.

**Comment by semorrison** [comment]
Otherwise, looks good.

File: src/Init/Data/String/Basic.lean (hunk 2 above)
Comment (austinletson) [reply src/Init/Data/String/Basic.lean]: I updated the single `"abc"` case. I agree this is more apparent.

I have streamlined the repetitive examples by replacing the "Given, `def...`" line with a let statement in the bullet:
```lean
`let lean := "L∃∀N"; lean.get' (0 |> lean.next |> lean.next) (by decide) = '∀'`
```

It seems off to me to inline the string in the repetitive cases: 
```lean
`"L∃∀N".get' (0 |> "L∃∀N".next |> "L∃∀N".next) (by decide) = '∀'`.
```

However, please let me know if you still prefer the inlining in the repetitive cases, and I will update the examples accordingly.

File: src/Init/Data/String/Basic.lean (hunk 1 above)
Comment (austinletson) [reply src/Init/Data/String/Basic.lean]: Updated

**Comment by austinletson** [comment]
awaiting-review

File: src/Init/Data/String/Basic.lean (hunk 2 above)
Comment (semorrison) [reply src/Init/Data/String/Basic.lean]: Looks good, thanks.

**Pull request merged**